File: engines/nunc-fluens/pipeline/src/topics.ts
// Topics: the single source of truth for what the pipeline searches for
// (topics-authoring plan W1/W2/W8). `data/reference/news-topics.json` holds
// user-authored topics with intent weights; the coverage gate, the compose
// fan-out, and the (future) authoring extractor all read it through here.
// The pre-plan `news-topics.md` converts lazily on first read (bullets →
// broad topics) — user data is never migrated by hand.
import { existsSync, readFileSync, writeFileSync } from 'node:fs';
import { join } from 'node:path';
import { REFERENCE_REL } from './world-paths.ts';

export type TopicIntent = 'deep' | 'broad' | 'watch';
export const TOPIC_INTENTS: TopicIntent[] = ['deep', 'broad', 'watch'];

export interface Topic {
  name: string;
  intent: TopicIntent;
  mandatory: boolean; // must be enumerated every run (orthogonal to intent)
  note?: string; // the user's natural-language trace; the gate shows it as a tag
}

export interface TopicsFile {
  topics: Topic[];
  // The md's "Default reference sites" section survives the migration here —
  // compose inlines them as prompt context, same as before.
  reference_sites?: string[];
}

export function topicsJsonPath(newsRepo: string): string {
  return join(newsRepo, ...REFERENCE_REL.split('/'), 'news-topics.json');
}

export function topicsMdPath(newsRepo: string): string {
  return join(newsRepo, ...REFERENCE_REL.split('/'), 'news-topics.md');
}

/** Strict parse: refuse rather than half-load an editorial-policy file. */
export function parseTopicsFile(raw: string): TopicsFile {
  let data: unknown;
  try {
    data = JSON.parse(raw);
  } catch {
    throw new Error('news-topics.json: not valid JSON');
  }
  if (typeof data !== 'object' || data === null || Array.isArray(data))
    throw new Error('news-topics.json: expected an object {topics, reference_sites?}');
  const obj = data as Record<string, unknown>;
  if (!Array.isArray(obj.topics) || obj.topics.length === 0)
    throw new Error('news-topics.json: topics must be a non-empty array');
  const seen = new Set<string>();
  const topics: Topic[] = obj.topics.map((t, i) => {
    if (typeof t !== 'object' || t === null) throw new Error(`news-topics.json: topics[${i}] is not an object`);
    const o = t as Record<string, unknown>;
    if (typeof o.name !== 'string' || !o.name.trim())
      throw new Error(`news-topics.json: topics[${i}].name must be a non-empty string`);
    if (!TOPIC_INTENTS.includes(o.intent as TopicIntent))
      throw new Error(`news-topics.json: topics[${i}].intent must be one of ${TOPIC_INTENTS.join('|')}`);
    if (typeof o.mandatory !== 'boolean')
      throw new Error(`news-topics.json: topics[${i}].mandatory must be a boolean`);
    if (o.note !== undefined && typeof o.note !== 'string')
      throw new Error(`news-topics.json: topics[${i}].note must be a string`);
    const name = o.name.trim();
    if (seen.has(name)) throw new Error(`news-topics.json: duplicate topic name ${JSON.stringify(name)}`);
    seen.add(name);
    const topic: Topic = { name, intent: o.intent as TopicIntent, mandatory: o.mandatory };
    if (typeof o.note === 'string' && o.note) topic.note = o.note;
    return topic;
  });
  const sites = Array.isArray(obj.reference_sites)
    ? obj.reference_sites.filter((s): s is string => typeof s === 'string' && !!s.trim())
    : undefined;
  return sites && sites.length ? { topics, reference_sites: sites } : { topics };
}

/**
 * One-time `.md` → `.json` conversion (plan migration note). Names are kept
 * VERBATIM — the coverage gate, verification.json entries, and the goldens
 * all match topic strings exactly. `mandatory` derives from the human's own
 * annotation in the name ("every run"); everything else lands `broad`.
 */
export function convertTopicsMd(md: string): TopicsFile {
  const listSection = /## Topic list\n([\s\S]*?)(?:\n## |$)/.exec(md)?.[1] ?? '';
  const topics: Topic[] = listSection
    .split('\n')
    .filter((l) => l.startsWith('- ') && !/For each of the above/i.test(l))
    .map((l) => l.slice(2).trim())
    .filter(Boolean)
    .map((name) => ({ name, intent: 'broad' as TopicIntent, mandatory: /every run/i.test(name) }));
  if (!topics.length) throw new Error('news-topics.md: no bullets under "## Topic list" to convert');
  const sitesSection = /## Default reference sites\n([\s\S]*?)(?:\n## |$)/.exec(md)?.[1] ?? '';
  const reference_sites = sitesSection
    .split('\n')
    .filter((l) => l.startsWith('- '))
    .map((l) => l.slice(2).trim())
    .filter(Boolean);
  return reference_sites.length ? { topics, reference_sites } : { topics };
}

export function serializeTopicsFile(tf: TopicsFile): string {
  return JSON.stringify(tf, null, 2) + '\n';
}

/**
 * Load an instance's topics: `.json` is the authority; a `.md` with no
 * `.json` beside it converts on first read and writes the `.json` back
 * (best-effort — a read-only view source still loads, it just converts
 * again next time). Neither file ⇒ an instructive refusal.
 */
export function loadTopics(newsRepo: string): TopicsFile {
  const jsonPath = topicsJsonPath(newsRepo);
  if (existsSync(jsonPath)) return parseTopicsFile(readFileSync(jsonPath, 'utf8'));
  const mdPath = topicsMdPath(newsRepo);
  if (existsSync(mdPath)) {
    const tf = convertTopicsMd(readFileSync(mdPath, 'utf8'));
    try {
      writeFileSync(jsonPath, serializeTopicsFile(tf));
      console.error(`topics: converted ${mdPath} -> news-topics.json (one-time migration)`);
    } catch {
      console.error(`topics: read-only source — converted ${mdPath} in memory only`);
    }
    return tf;
  }
  throw new Error(
    `no news-topics.json (or legacy news-topics.md) under ${newsRepo}/${REFERENCE_REL} — `
      + 'author topics in the Formans world view, or seed the file from the instance template',
  );
}

export const topicNames = (tf: TopicsFile): string[] => tf.topics.map((t) => t.name);

/**
 * The compose prompt block — shaped like the old news-topics.md so the
 * writer prompt stays stable across the migration.
 */
export function topicsPromptBlock(tf: TopicsFile): string {
  const lines = ['## Topic list', '', ...tf.topics.map((t) => `- ${t.name}`)];
  if (tf.reference_sites?.length) {
    lines.push('', '## Default reference sites', '', ...tf.reference_sites.map((s) => `- ${s}`));
  }
  return lines.join('\n') + '\n';
}

File: gate/src/topics.rs
//! Topic authoring API (topics-authoring W6): the gate is the one door, so
//! the Formans World panel reads and writes the LINKED instance's
//! `data/reference/news-topics.json` through it — the single topic
//! authority the pipeline's coverage gate and compose fan-out both consume
//! (src/topics.ts). Same discipline as news_config.rs: deny-unknown fields,
//! validate, atomic tmp+rename.
//!
//! W7 read-only guard: edits target the linked instance, and PUT is refused
//! (409) unless that instance is a writable init-born one (has an
//! `instance.json` stamp). A news-shaped checkout or a bare view source is
//! never written (the fork rule — never write production ~/news).

use std::path::{Path, PathBuf};

use axum::{
    Json,
    extract::State,
    http::StatusCode,
    response::{IntoResponse, Response},
};
use serde::{Deserialize, Serialize};

use crate::GateCfg;

/// One authored topic. Mirrors src/topics.ts `Topic`.
#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(deny_unknown_fields)]
pub struct Topic {
    pub name: String,
    /// The search-weight axis: deep | broad | watch.
    pub intent: String,
    /// Must be enumerated every run regardless of intent (orthogonal).
    pub mandatory: bool,
    /// The user's natural-language trace; renders as the gate's tag.
    #[serde(skip_serializing_if = "Option::is_none")]
    pub note: Option<String>,
}

/// The file shape. Mirrors src/topics.ts `TopicsFile`.
#[derive(Debug, Default, Clone, Serialize, Deserialize)]
#[serde(deny_unknown_fields)]
pub struct TopicsFile {
    pub topics: Vec<Topic>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub reference_sites: Option<Vec<String>>,
}

/// GET wraps the file with the writability the UI needs to show a
/// read-only banner (the file itself never carries it).
#[derive(Debug, Serialize)]
pub struct TopicsResponse {
    pub topics: Vec<Topic>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub reference_sites: Option<Vec<String>>,
    pub writable: bool,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub read_only_reason: Option<String>,
}

impl TopicsFile {
    fn validate(&self) -> Result<(), String> {
        if self.topics.is_empty() {
            return Err("topics must be a non-empty array".into());
        }
        let mut seen = std::collections::HashSet::new();
        for (i, t) in self.topics.iter().enumerate() {
            if t.name.trim().is_empty() {
                return Err(format!("topics[{i}].name must not be empty"));
            }
            if !matches!(t.intent.as_str(), "deep" | "broad" | "watch") {
                return Err(format!(
                    "topics[{i}].intent must be one of deep|broad|watch, got {:?}",
                    t.intent
                ));
            }
            if !seen.insert(t.name.trim()) {
                return Err(format!("duplicate topic name {:?}", t.name.trim()));
            }
        }
        Ok(())
    }
}

fn topics_path(news_repo: &Path) -> PathBuf {
    news_repo
        .join("data")
        .join("reference")
        .join("news-topics.json")
}

/// W7: an init-born instance carries an `instance.json` stamp; a view
/// source / news-shaped checkout does not, and must never be written.
fn is_writable_instance(news_repo: &Path) -> bool {
    news_repo.join("instance.json").is_file()
}

fn no_instance() -> Response {
    (
        StatusCode::SERVICE_UNAVAILABLE,
        "no nunc-fluens instance linked - point the world view at one with \
         `just news-link <instance>` (create one with `just news-init`)",
    )
        .into_response()
}

pub async fn get_topics(State(cfg): State<GateCfg>) -> Response {
    let Some(news_repo) = &cfg.news_repo else {
        return no_instance();
    };
    if !news_repo.exists() {
        return (
            StatusCode::NOT_FOUND,
            format!(
                "linked instance {} does not exist - re-link with `just news-link`",
                news_repo.display()
            ),
        )
            .into_response();
    }
    let writable = is_writable_instance(news_repo);
    let read_only_reason = (!writable).then(|| {
        "the linked source is not an init-born instance (no instance.json) - \
         it is a read-only view source; author on a writable instance created \
         with `just news-init`"
            .to_string()
    });

    let path = topics_path(news_repo);
    let file = match tokio::fs::read(&path).await {
        Ok(bytes) => match serde_json::from_slice::<TopicsFile>(&bytes) {
            Ok(parsed) => parsed,
            Err(e) => {
                return (
                    StatusCode::INTERNAL_SERVER_ERROR,
                    format!("news-topics.json is unreadable: {e}"),
                )
                    .into_response();
            }
        },
        // Absent = not yet migrated from .md, or a fresh authoring surface.
        // The UI authors from an empty list (the pipeline migrates .md on
        // its own next run; we never convert in Rust).
        Err(_) => TopicsFile::default(),
    };

    Json(TopicsResponse {
        topics: file.topics,
        reference_sites: file.reference_sites,
        writable,
        read_only_reason,
    })
    .into_response()
}

pub async fn put_topics(
    State(cfg): State<GateCfg>,
    payload: Result<Json<TopicsFile>, axum::extract::rejection::JsonRejection>,
) -> Response {
    let Some(news_repo) = &cfg.news_repo else {
        return no_instance();
    };
    if !is_writable_instance(news_repo) {
        return (
            StatusCode::CONFLICT,
            "the linked source is read-only (not an init-born instance) - topics \
             can only be authored on a writable instance created with `just news-init`",
        )
            .into_response();
    }
    let Json(next) = match payload {
        Ok(p) => p,
        Err(e) => return (StatusCode::BAD_REQUEST, format!("invalid body: {e}")).into_response(),
    };
    if let Err(msg) = next.validate() {
        return (StatusCode::UNPROCESSABLE_ENTITY, msg).into_response();
    }

    let path = topics_path(news_repo);
    if let Some(parent) = path.parent() {
        if let Err(e) = tokio::fs::create_dir_all(parent).await {
            return (
                StatusCode::INTERNAL_SERVER_ERROR,
                format!("cannot create {}: {e}", parent.display()),
            )
                .into_response();
        }
    }
    // Atomic write: tmp + rename (the news_config.rs / pipeline discipline).
    let tmp = path.with_extension("json.tmp");
    let body = match serde_json::to_vec_pretty(&next) {
        Ok(mut v) => {
            v.push(b'\n');
            v
        }
        Err(e) => {
            return (StatusCode::INTERNAL_SERVER_ERROR, format!("serialize: {e}")).into_response();
        }
    };
    if let Err(e) = tokio::fs::write(&tmp, &body).await {
        return (
            StatusCode::INTERNAL_SERVER_ERROR,
            format!("write {}: {e}", tmp.display()),
        )
            .into_response();
    }
    if let Err(e) = tokio::fs::rename(&tmp, &path).await {
        let _ = tokio::fs::remove_file(&tmp).await;
        return (
            StatusCode::INTERNAL_SERVER_ERROR,
            format!("rename into place: {e}"),
        )
            .into_response();
    }
    Json(next).into_response()
}

// --- NL structuring (W4): thin proxy to the local model -------------------
// The browser sends a plain-language request; the gate asks the local
// `llama-server` (OpenAI-compatible) to structure it into the topic model,
// constrained by a json_schema, and returns the model's JSON. The browser
// merges the proposal into the existing set and shows the diff before
// saving (W5/F6 — nothing is written unattended). The gate holds NO topic
// logic beyond the shape it already knows; prompt wording stays minimal.
// Model down ⇒ 503, and the UI degrades to "run just setup".

#[derive(Deserialize)]
#[serde(deny_unknown_fields)]
pub struct ExtractRequest {
    /// The user's natural-language request.
    request: String,
    /// Existing topic names, so the model refines rather than duplicates.
    #[serde(default)]
    existing: Vec<String>,
}

pub async fn extract_topics(
    State(cfg): State<GateCfg>,
    payload: Result<Json<ExtractRequest>, axum::extract::rejection::JsonRejection>,
) -> Response {
    let Json(req) = match payload {
        Ok(p) => p,
        Err(e) => return (StatusCode::BAD_REQUEST, format!("invalid body: {e}")).into_response(),
    };
    if req.request.trim().is_empty() {
        return (StatusCode::BAD_REQUEST, "request must not be empty").into_response();
    }

    // The topic model as a json_schema, so the model's output is constrained.
    let schema = serde_json::json!({
        "type": "object",
        "properties": {
            "topics": {
                "type": "array",
                "items": {
                    "type": "object",
                    "properties": {
                        "name": {"type": "string"},
                        "intent": {"type": "string", "enum": ["deep", "broad", "watch"]},
                        "mandatory": {"type": "boolean"},
                        "note": {"type": "string"}
                    },
                    "required": ["name", "intent", "mandatory"]
                }
            }
        },
        "required": ["topics"]
    });
    let system = format!(
        "You structure a user's request into research topics for a news pipeline. \
         Output ONLY JSON matching the schema: an array of {{name, intent, mandatory, note}}. \
         intent: deep = a topic to investigate thoroughly, broad = a light multi-angle scan, \
         watch = a light single check. mandatory = must be covered every run. \
         Reuse these existing topic names verbatim when the request refines them: [{}].",
        req.existing.join(", ")
    );
    let body = serde_json::json!({
        "messages": [
            {"role": "system", "content": system},
            {"role": "user", "content": req.request}
        ],
        "response_format": {"type": "json_schema", "json_schema": {"name": "topics", "schema": schema, "strict": true}},
        "stream": false,
        "temperature": 0.2
    });

    let url = format!("{}/v1/chat/completions", cfg.llama_url.trim_end_matches('/'));
    let res = match cfg
        .client
        .post(&url)
        .json(&body)
        .timeout(std::time::Duration::from_secs(120))
        .send()
        .await
    {
        Ok(r) => r,
        // Connection refused / DNS / timeout = the model is not running.
        Err(_) => {
            return (
                StatusCode::SERVICE_UNAVAILABLE,
                "local model offline - run `just setup` to enable AI structuring \
                 (manual topic editing works without it)",
            )
                .into_response();
        }
    };
    if !res.status().is_success() {
        let code = res.status();
        let detail = res.text().await.unwrap_or_default();
        return (
            StatusCode::BAD_GATEWAY,
            format!("local model error ({code}): {}", detail.chars().take(300).collect::<String>()),
        )
            .into_response();
    }
    let completion: serde_json::Value = match res.json().await {
        Ok(v) => v,
        Err(e) => return (StatusCode::BAD_GATEWAY, format!("model reply unreadable: {e}")).into_response(),
    };
    // OpenAI shape: choices[0].message.content is the JSON string.
    let content = completion["choices"][0]["message"]["content"]
        .as_str()
        .unwrap_or("");
    match serde_json::from_str::<serde_json::Value>(content) {
        Ok(parsed) => Json(parsed).into_response(),
        Err(e) => (
            StatusCode::BAD_GATEWAY,
            format!("model did not return valid topic JSON: {e}"),
        )
            .into_response(),
    }
}
